### PR TITLE
Use of "iterable" Type May Need Re-Evaluation Across the Repo 

### PR DESCRIPTION
**Pull request by amber0612**

### CODE REVIEW

File: tests/Unit/SDK/Common/Util/FunctionsTest.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        $this->assertTrue(isEmpty([]));
+    }
+
+    public function test_for_exhausted_generator(): void
+    {
+        $generator = gen();
+        foreach ($generator as $value) {
+        }
+        $value=isEmpty($generator) ? true : false;
```
Comment (Nevay) [inline tests/Unit/SDK/Common/Util/FunctionsTest.php:30]: Obsolete ternary operator

**Review by Nevay — COMMENTED** [review]
Failing test:
```php
assert(isEmpty((fn() => yield from [])()));
```

File: tests/Unit/SDK/Common/Util/FunctionsTest.php (hunk 1 above)
Comment (amber0612) [reply tests/Unit/SDK/Common/Util/FunctionsTest.php]: Resolved

**Comment by Nevay** [comment]
Another failing test:
```php
$generator = (fn() => yield from [1, 2])();
$generator->next();
assert(!isEmpty($generator));
```

**Comment by amber0612** [comment]
$generator = (fn() => yield from [1, 2])();
$generator->next();
assert(!isEmpty($generator));

This is a non empty generator case and hence isEmpty() will return false. Hence the test case is passing

**Comment by Nevay** [comment]
> isEmpty() will return false

Currently it returns true (`foreach` tries to rewind the generator and triggers an exception).

---

What is the expected result for an exhausted rewindable iterator?
```php
$iterator = new ArrayIterator([1]);
$iterator->next();
$empty = isEmpty($iterator); // ?
```

Should `isEmpty()` be allowed to modify the given iterable (somewhat obsolete if we mark the function as internal)?
```php
$iterator = new ArrayIterator([1, 2]);
$iterator->next();
isEmpty($iterator);
$value = $iterator->current(); // ?
```

Side note: We could simply remove the [special handling of an empty batch](https://github.com/open-telemetry/opentelemetry-php/blob/bf912ed7bd58a7a9bfb034bcf2907b2789bdb4e2/src/SDK/Trace/Behavior/SpanExporterTrait.php#L44-L46) and the `isEmpty()` function, our span processors call `SpanExporter::export()` with a non-empty batch / exporters can skip empty exports after conversion.

**Comment by amber0612** [comment]
> 
Makes sense. The iterable will be modified.

**Review by Nevay — APPROVED** [review]

**Review by brettmc — APPROVED** [review]

**Pull request merged**